Add comprehensive horizontal pod autoscaling documentation

itskingori opened the pull request:
As requested by @chrislovecnm in https://github.com/kubernetes/kops/pull/3939#pullrequestreview-79327244.

Please note that this is a first pass. I'm currently trying to set up HPA and these are the things I've had to set. I'm yet to actually roll out the changes to the cluster and so I might update these docs with new discoveries.

zacblazic commented on docs/horizontal_pod_autoscaling.md:
````diff
@@ -0,0 +1,85 @@
+  --proxy-client-cert-file=<path to aggregator proxy cert>
+  --proxy-client-key-file=<path to aggregator proxy key>
+  ```
+* [kubernetes/kops#3679][4] - add config option to set `--horizontal-pod-autoscaler-use-rest-clients` kube-controller-manager flag required to enable custom metrics
````
The link name `kubernetes/kops#3679` should be `kubernetes/kops#3939`.

itskingori replied:
Fixed in https://github.com/kubernetes/kops/pull/3942/commits/762589845644d385d219fef495248c21a3834abb.

zacblazic commented on docs/horizontal_pod_autoscaling.md:
```diff
@@ -0,0 +1,85 @@
+Kops can assist in setting up HPA and recommends Kubernetes `>= 1.7` and Kops
+`>=1.8`. Relevant reading you will need to go through:
+
+* [Configure The Aggregation Layer][1].
```
No periods at the end looks nicer IMO.

itskingori replied:
Fixed in https://github.com/kubernetes/kops/pull/3942/commits/7772c448fe66d1cc0d68e6e8e6245179311ffb20.

mikesplain commented:
If we're breaking out HPA docs, can we link to it from here: https://github.com/kubernetes/kops/blob/master/docs/cluster_spec.md#kubecontrollermanager

chrislovecnm commented:
/ok-to-test
/lgtm

itskingori commented:
@mikesplain yeah, there's some duplication 😓 ... both docs makes sense. I think your suggestion makes sense.

chrislovecnm commented:
/lgtm cancel

chrislovecnm commented:
Thinking about something. Does this feature use the api aggregation server? I know we have some nasty bugs in 1.8 that are being worked on

itskingori commented:
> If we're breaking out HPA docs, can we link to it ...

@mikesplain addressed in https://github.com/kubernetes/kops/pull/3942/commits/c38aa8be986046889f6ea9fe1dc5d87af67d3dba.

@chrislovecnm sure, no problem. We can keep this open for discussion until everything has been resolved.

chrislovecnm commented:
@itskingori  should we merge with a note about the bug, or just keep this open?  thoughts?  I am leaning toward keeping this open.

chrislovecnm commented:
/hold

itskingori commented:
@chrislovecnm we can keep it open. I'm currently setting up HPA on 1.7 and can keep updating this as I roll it out).

23doors commented:
Any progress on this? What about custom metrics aggregator https://github.com/kubernetes/metrics/blob/master/IMPLEMENTATIONS.md ? Guess it still needs to be deployed.

mikesplain commented:
@chrislovecnm Any updates on this? I think it'd be good to get these docs in eventually.  I believe the 1.8 bugs have since been resolved and this could be a good starting place for future HPA docs.  Thoughts?

mikesplain commented:
@itskingori any other recommended updates since you last touched this?

/assign

itskingori commented:
@mikesplain Oh my, forgot about this. Learnt quite a bit, there's quite a bit to add. Will make some time tomorrow.

mikesplain commented:
Ping @itskingori :) Would love to get the docs updated with anything you've learned for 1.9!

mikesplain commented:
Oh that's why I got pinged, nevermind!

itskingori commented:
@mikesplain yeah 😅 ... added some notes. I'm kinda torn in the sense that I don't want to regurgitate what's in the docs. So I've made mention of everything that's needed i.e. all the necessary configuration (from an implementation point of view), the cluster config that one needs and the metrics servers (resource and custom) that need to the registered (needs `kubectl apply` only).

I'm wondering if having the manifest files for the metrics servers in kops as addons is a good idea or just send people to the relevant repos/docs. 🤔

itskingori commented:
> I'm wondering if having the manifest files for the metrics servers in kops as addons is a good idea or just send people to the relevant repos/docs. 🤔 

@mikesplain about this ... decided to create a separate PR, see https://github.com/kubernetes/kops/pull/4581.

mikesplain commented:
This is great @itskingori! Thanks for all the detail! I agree that breaking that into a separate PR is a good call. 

cc: @chrislovecnm I think this looks good now if you want to take another look and consider removing the hold?

/area documentation
/lgtm

itskingori commented:
@chrislovecnm I never quite responded to your suggestion in https://github.com/kubernetes/kops/pull/3942#issuecomment-347600816. I've addressed it in https://github.com/kubernetes/kops/pull/3942/commits/929f299faca266e1d71aef00af1facfba90670f8. @mikesplain added one last commit 🙈 post-LGTM.

@23doors about https://github.com/kubernetes/kops/pull/3942#issuecomment-355834622, addressed in https://github.com/kubernetes/kops/pull/3942/commits/472604cac3f90abfbb82a3e4fd9c903c017f9e27.  See also https://github.com/kubernetes/kops/pull/4581.

mikesplain commented:
Awesome, good fix!

/lgtm

chrislovecnm commented:
/lgtm
